src/editable_tree/mod.rs

```diff
@@ -584,45 +585,1164 @@ impl<'arena, Node: Ast<'arena>> DAG<'arena, Node> {
 }
 
 #[cfg(test)]
-mod integration {
-    use super::{cursor_path::CursorPath, EditSuccess, DAG};
+mod tests {
+    use super::{cursor_path::CursorPath, Direction, EditErr, EditSuccess, Side, DAG};
     use crate::arena::Arena;
-    use crate::ast::test_json::TestJSON as J;
+    use crate::ast::json::JSON;
+    use crate::ast::test_json::TestJSON;
     use crate::editor::normal_mode::Action;
 
+    fn run_test(
+        start_tree: TestJSON,
+        start_cursor_location: CursorPath,
+        action: Action,
+        expected_edit_result: (bool, Result<EditSuccess, EditErr>),
+        expected_tree: TestJSON,
+        expected_cursor_location: CursorPath,
+    ) {
+        let arena: Arena<JSON> = Arena::new();
+        let root = start_tree.add_to_arena(&arena);
+        let mut editable_tree = DAG::new(&arena, root, start_cursor_location);
+
+        assert_eq!(
+            expected_edit_result,
+            editable_tree.execute_action(action),
+            "Not equal in action result"
+        );
+        assert_eq!(expected_tree, editable_tree.root(), "Not equal in tree.");
+        assert_eq!(
+            expected_cursor_location, editable_tree.current_cursor_path,
+            "Not equal in cursor location."
+        );
+    }
+
+    #[test]
+    fn dag_root_insertchild() {
+        let start_tree = TestJSON::Array(vec![]);
+        let start_cursor_location = CursorPath::from_vec(vec![]);
```

Comment:
There's a special method to refer to the root of the tree: `CursorPath::root()` is the same as `CursorPath::from_vec(vec![])`, and is much clearer.  You could probably just do a file-wide find/replace for this, since `CursorPath::from_vec(vec![])` is literally how `CursorPath::root()` is defined, so they are completely interchangable.